plugins/tactics/src/Ide/Plugin/Tactic/CodeGen.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE FlexibleContexts #-}
+module Ide.Plugin.Tactic.CodeGen where
+
+import Control.Monad.Except
+import Data.List
+import Data.Traversable
+import DataCon
+import Development.IDE.GHC.Compat
+import GHC.Exts
+import GHC.SourceGen.Binds
+import GHC.SourceGen.Expr
+import GHC.SourceGen.Overloaded
+import GHC.SourceGen.Pat
+import Ide.Plugin.Tactic.Judgements
+import Ide.Plugin.Tactic.Machinery
+import Ide.Plugin.Tactic.Naming
+import Ide.Plugin.Tactic.Types
+import Name
+import Type hiding (Var)
+
+
+destructMatches
+    :: (DataCon -> Judgement -> Rule)
+       -- ^ How to construct each match
+    -> (Judgement -> Judgement)
+       -- ^ How to derive each match judgement
+    -> CType
+       -- ^ Type being destructed
+    -> Judgement
+    -> RuleM [RawMatch]
+destructMatches f f2 t jdg = do
+  let hy = jHypothesis jdg
+      g  = jGoal jdg
+  case splitTyConApp_maybe $ unCType t of
+    Nothing -> throwError $ GoalMismatch "destruct" g
+    Just (tc, apps) -> do
+      let dcs = tyConDataCons tc
```

Comment:
Are these `DataCons` in scope? Are they even known to be exported?

Reply:
Neither. I don't think the first is important; other plugins can suggest the import.

That being said, preventing unexported datacons from being generated is more interesting, Is there an easy way we can check that?